src/contact.rs

```diff
@@ -189,6 +217,134 @@ pub async fn make_vcard(context: &Context, contacts: &[ContactId]) -> Result<Str
     Ok(contact_tools::make_vcard(&vcard_contacts))
 }
 
+/// Imports contacts from the given vCard.
+///
+/// Returns the ids of created/modified contacts in the order they appear in `vcard`.
+pub async fn import_vcard(
+    context: &Context,
+    vcard: &str,
+    timestamp: i64,
+) -> Result<Vec<ContactId>> {
+    let contacts = contact_tools::parse_vcard(vcard);
+    let mut contact_ids = Vec::with_capacity(contacts.len());
+    for c in &contacts {
+        let Ok(Some(id)) = import_vcard_contact(context, c, timestamp)
+            .await
+            .with_context(|| format!("import_vcard_contact() failed for {}", c.addr))
+            .log_err(context)
+        else {
+            continue;
+        };
+        contact_ids.push(id);
+    }
+    Ok(contact_ids)
+}
+
+async fn import_vcard_contact(
+    context: &Context,
+    contact: &VcardContact,
+    timestamp: i64,
+) -> Result<Option<ContactId>> {
+    let addr = ContactAddress::new(&contact.addr).context("Invalid address")?;
+    // Importing a vCard is also an explicit user action like creating a chat with the contact. We
+    // mustn't use `Origin::AddressBook` here because the vCard may be created not by us, also we
+    // want `contact.authname` to be saved as the authname and not a locally given name.
+    let origin = Origin::CreateChat;
+    let (id, modified) =
+        match Contact::add_or_lookup(context, &contact.authname, &addr, origin).await {
+            Err(e) => return Err(e).context("Contact::add_or_lookup() failed"),
+            Ok((ContactId::SELF, _)) => return Ok(None),
+            Ok(val) => val,
+        };
+    if modified != Modifier::None {
+        context.emit_event(EventType::ContactsChanged(Some(id)));
+    }
+    let key = contact.key.as_ref().and_then(|k| {
+        SignedPublicKey::from_base64(k)
+            .with_context(|| {
+                format!(
+                    "import_vcard_contact: Cannot decode key for {}",
+                    contact.addr
+                )
+            })
+            .log_err(context)
+            .ok()
+    });
+    if let Some(public_key) = key {
+        let timestamp = *contact.timestamp.as_ref().unwrap_or(&timestamp);
```

Comment:
Something I just noticed: We need to make sure that `contact.timestamp` is not bigger than the current time. Right now, if we click on a vCard with a timestamp far in the future, we may set the last_seen to a very big number in the db and reject future updates. I.e. in order to combine it with @link2xt's comment about not using mtime, this line becomes:

```rust
let current_time = smeared_time(context)
let timestamp = timestamp
            .map_or(current_time, |value| min(value, current_time));
```

Reply:
Totally agree! The only thing is that i decided to use 0 as the default timestamp (see the comment above)